### ISSUE TITLE
File contents duplicated on save after multi-line input

### ISSUE THREAD

**Issue by andypearson** [glennr/uber-glory-tmbundle#10]
This is a bit of an odd one, steps to reproduce as follows:
1. Create a blank file, and save it somewhere
2. In that file enter the following text:
   
   One
   Two
   Three
3. Using multi-line select (hold alt) change the text to:
   If you do it correctly, you only need 1 key-stroke to edit all 3 lines
   
   -One
   -Two
   -Three
4. Hit Command+S to save the file.

The contents of the file should now be:

```
-One
-Two
-Three-One
-Two
-Three-One
-Two
-Three
```

**Comment by gavinlaking** [comment]
Also experiencing this, I thought it was something to do with copy and paste and wasn't entirely sure what I was doing to provoke this bug.

**Comment by andypearson** [comment]
Yes, it took me a while to realise what was going on!

**Comment by skipchris** [comment]
I've forked and created an experimental fix for this issue here:

https://github.com/skipchris/uber-glory-tmbundle

It seems to work fine for me, but I'm not an expert in TM bundles, and I think I've removed the functionality to strip whitespace from only a selection (which I'm fine with as I never want to do that), so YMMV.

**Comment by andypearson** [comment]
@skipchris Nice, I'll give it a try as soon as I can and let you know how it goes! Merry Christmas!

**Comment by andypearson** [comment]
@skipchris Finally had a chance to look at your fork, from what I can tell it actually makes the duplication quite a lot worse! I can't really work out what changes you made (the diff contains most of the file).

**Comment by skipchris** [comment]
Hey Andy,

Apologies for the dodgy diff - I think my re-save changed tabs-spaces or vice-versa (probably by having UberGlory installed!). :/

The change I made was getting rid of the 'fallbackInput' key/value pair, and changing 'input' to be 'document' (which was formerly set to 'selection') - my logic being that it was having a 'column' selection which was confusing TM (or perl), and I wasn't using the 'selection' input mode at all.

Working tons better for me, sorry it hasn't fixed things for you - I guess my only other suggestion is we keep our fingers crossed for TM2...

**Comment by andypearson** [comment]
@skipchris I tried manually applying the fix (and checking it in the bundle editor) but even after that I still see the same behaviour. Hopefully @glennr will chime in at some point...

Happy New Year!

**Comment by glennr** [comment]
Happy new year guys

Just reproduced this one locally. 

No idea why it's dumping out duplicated alt-selected-blocks. The Uber-Glory regexp itself is quite simple, and probably not the source of the issue. Like Chris, I suspect it has more to do with the combination of alt-selected blocks and these options that we use in UG;
    input = selection;
    output = replaceSelectedText;

Chris - can you make a patch with your fix? (i.e. input = document, I assume?). 

I'd need to check the TM API again re: selections, but intuitively it makes more sense to ingore selections completely, and always apply the regexp to the entire document. After all, CMD+S operates on the entire document...

**Comment by andypearson** [comment]
@glennr it seems to be that it is a selection problem, as in, it seems to duplicate for the number of lines that are alt-selected. You basically need `output = replaceDocument;` but I have no idea if that is a thing.

**Comment by skipchris** [comment]
Just a note to say that my patch is indeed buggered - it seemed to improve things for a while, but perhaps I was testing under different conditions, because the problem's back with a vengeance.

My solution has been to switch to Sublime Text 2 - it has an in-built config flag to strip trailing whitespace on save. :)

**Comment by aseemk** [comment]
Do people use/like the feature where, if text is selected, only the selected text is processed? Might it fix this bug and simplify the command if it just always acted on the entire document?

**Comment by andypearson** [comment]
@aseemk - I didn't even know that was part of the feature set! And to be honest, even if you act on selected text, you are just going to process the whole file when you hit command-S to save the document anyway...

**Comment by aseemk** [comment]
Yep, I agree — just wanted to check if others used it. I think it'd be reasonable to (edit: always) act on the whole document if @glennr and others do too. =)

**Comment by joshferrara** [comment]
+1 on this issue. I'm not sure what the general consensus is, but I wouldn't mind if a fix required the removal of selective processing!

**Comment by kayzee** [comment]
I agree as well.  I often have text selected and hit CMD+S.  It always adds a line break, which is not noticed, can cause serious issues with certain file types (haml comes to mind).

Please have the plugin act on the full document.
